common/authorization/interceptor.go

```diff
@@ -89,8 +89,10 @@ func (a *interceptor) Interceptor(
 			tlsSubject = &clientCert.Subject
 		}
 
+		_, authInfoNotRequired := a.claimMapper.(interface{ AuthInfoNotRequired() })
```

Comment:
This is a really unintuitive way of determining whether a method supports a nil argument. I think either of these two are better, but I heavily favor option 2.

1 ) Add a `AuthInfoRequired() bool` method to the claim mapper interface
2 ) Pass nil auth info into the GetClaims method and let implementations decide whether that's an error or not.

Reply:
Totally agreed that those are cleaner, but I don't think we can do either one of them:
1. That would break the build for existing claim mappers that don't have that new method. It's an easy fix, but we have promised compatibility here.
2. Since GetClaims has never been called with a nil argument before, I imagine that some implementations have been written to assume that it's non-nil, so this could make them crash.

Reply:
Oh, I assumed that this was an interface with just two internal implementations. Looking at [WithClaimMapper](https://github.com/temporalio/temporal/blob/0b4df6a678544a1e214ebe6c25bfbd3548c252a5/temporal/server_option.go#L116), I see that this is an interface that we expect users to sometimes implement and provide. In that case, it makes sense that we need to workaround the fact that we can't add methods to the interface.

However, I think that this should still be `AuthInfoRequired() bool`, and we should check the return value, so that
a) it doesn't look as awkward on the side of the implementor
b) if we do decide to make this a required method later, breaking users that haven't implemented it, it at least won't break those who already have implemented it

Reply:
Actually I was wrong about the nil pointer, even with no auth info this will pass a non-nil pointer. But even so, I'd rather be really conservative with the semantics here and only change behavior if opted in, since this is a security-critical area

Reply:
Yeah, I agree